# Import Package

In [1]:
import pandas as pd
import numpy as np
import datetime as dt

# Load the Data

In [2]:
df = pd.read_excel('Online Retail Data.xlsx', header=0)
df

order_id product_code                     product_name  quantity  \
0       493410.0      TEST001          This is a test product.       5.0   
1        C493411      21539.0          RETRO SPOTS BUTTER DISH      -1.0   
2       493412.0      TEST001          This is a test product.       5.0   
3       493413.0      21724.0  PANDA AND BUNNIES STICKER SHEET       1.0   
4       493413.0      84578.0   ELEPHANT TOY WITH BLUE T-SHIRT       1.0   
...          ...          ...                              ...       ...   
461768  539991.0      21618.0   4 WILDFLOWER BOTANICAL CANDLES       1.0   
461769  539991.0      72741.0            GRAND CHOCOLATECANDLE       4.0   
461770  539992.0      21470.0    FLOWER VINE RAFFIA FOOD COVER       1.0   
461771  539992.0      22258.0          FELT FARM ANIMAL RABBIT       1.0   
461772  539992.0      21155.0            RED RETROSPOT PEG BAG       1.0   

                order_date  price  customer_id  
0      2010-01-04 09:24:00   4.50      12346.0  
1      2010-01-04 09:43:00   4.25      14590.0  
2      2010-01-04 09:53:00   4.50      12346.0  
3      2010-01-04 09:54:00   0.85          NaN  
4      2010-01-04 09:54:00   3.75          NaN  
...                    ...    ...          ...  
461768 2010-12-23 16:49:00   1.25          NaN  
461769 2010-12-23 16:49:00   1.45          NaN  
461770 2010-12-23 17:41:00   3.75          NaN  
461771 2010-12-23 17:41:00   1.25          NaN  
461772 2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      461773 non-null  object        
 1   product_code  461773 non-null  object        
 2   product_name  459055 non-null  object        
 3   quantity      461773 non-null  float64       
 4   order_date    461773 non-null  datetime64[ns]
 5   price         461773 non-null  float64       
 6   customer_id   360853 non-null  float64       
dtypes: datetime64[ns](1), float64(3), object(3)
memory usage: 24.7+ MB


# Data Cleansing

In [4]:
df_clean = df.copy()
# Create a date column
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64')
# Remove all rows without customer_id
df_clean = df_clean[~df_clean['customer_id'].isna()]
# Remove all rows without product_name
df_clean = df_clean[~df_clean['product_name'].isna()]
# Convert customer_id, product_code, product_name to string
df_clean['customer_id'] = df_clean['customer_id'].astype(str)
df_clean['product_code'] = df_clean['product_code'].astype(str)
df_clean['product_name'] = df_clean['product_name'].astype(str)
# Convert all product_name to lowercase
df_clean['product_name'] = df_clean['product_name'].str.lower()
# Remove all rows with product_code or product_name containing 'test'
df_clean = df_clean[(~df_clean['product_code'].str.lower().str.contains('test')) |
                    (~df_clean['product_name'].str.contains('test '))]
# Create an order_status column with 'cancelled' value if order_id starts with 'c', and 'delivered' otherwise
df_clean['order_status'] = np.where(df_clean['order_id'].str[:1]=='C', 'cancelled', 'delivered')
# Convert negative quantity values to positive since negative values indicate cancelled orders
df_clean['quantity'] = df_clean['quantity'].abs()
# Remove rows with negative price
df_clean = df_clean[df_clean['price']>0]
# Create amount value, which is the multiplication of quantity and price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']
# Replace product_name from product_code that has multiple product_names with the most frequently occurring one
most_freq_product_name = df_clean.groupby(['product_code','product_name'], as_index=False).agg(order_cnt=('order_id','nunique')).sort_values(['product_code','order_cnt'], ascending=[True,False])
most_freq_product_name['rank'] = most_freq_product_name.groupby('product_code')['order_cnt'].rank(method='first', ascending=False)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank']==1].drop(columns=['order_cnt','rank'])
df_clean = df_clean.merge(most_freq_product_name.rename(columns={'product_name':'most_freq_product_name'}), how='left', on='product_code')
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')
# Remove outliers
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean

order_id product_code                       product_name  quantity  \
0        C493411      21539.0          red retrospot butter dish       1.0   
1       493414.0      21844.0                  red retrospot mug      36.0   
2       493414.0      21533.0          retro spot large milk jug      12.0   
3       493414.0      37508.0    new england ceramic cake server       2.0   
4       493414.0       35001G               hand open shape gold       2.0   
...          ...          ...                                ...       ...   
358464  539988.0      84380.0  set of 3 butterfly cookie cutters       1.0   
358465  539988.0       84849D              hot baths soap holder       1.0   
358466  539988.0       84849B             fairy soap soap holder       1.0   
358467  539988.0      22854.0        cream sweetheart egg holder       2.0   
358468  539988.0       47559B                tea time oven glove       2.0   

                order_date  price customer_id       date order_status  amount  
0      2010-01-04 09:43:00   4.25     14590.0 2010-01-04    cancelled    4.25  
1      2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered   91.80  
2      2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered   51.00  
3      2010-01-04 10:28:00   2.55     14590.0 2010-01-04    delivered    5.10  
4      2010-01-04 10:28:00   4.25     14590.0 2010-01-04    delivered    8.50  
...                    ...    ...         ...        ...          ...     ...  
358464 2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    1.25  
358465 2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358466 2010-12-23 16:06:00   1.69     18116.0 2010-12-23    delivered    1.69  
358467 2010-12-23 16:06:00   4.95     18116.0 2010-12-23    delivered    9.90  
358468 2010-12-23 16:06:00   1.25     18116.0 2010-12-23    delivered    2.50  

[358469 rows x 10 columns]

In [5]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358469 entries, 0 to 358468
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      358469 non-null  object        
 1   product_code  358469 non-null  object        
 2   product_name  358469 non-null  object        
 3   quantity      358469 non-null  float64       
 4   order_date    358469 non-null  datetime64[ns]
 5   price         358469 non-null  float64       
 6   customer_id   358469 non-null  object        
 7   date          358469 non-null  datetime64[ns]
 8   order_status  358469 non-null  object        
 9   amount        358469 non-null  float64       
dtypes: datetime64[ns](2), float64(3), object(5)
memory usage: 27.3+ MB


# Creating RFM Segmentation

### Aggregate transaction data into a summary of total transactions (orders), total order value, and the last order date for each user.

In [6]:
df_user = df_clean.groupby('customer_id', as_index=False).agg(order_cnt=('order_id', 'nunique'), max_order_date=('date','max'), total_order_value=('amount','sum'))
df_user

customer_id  order_cnt max_order_date  total_order_value
0        12346.0          5     2010-10-04             602.40
1        12608.0          1     2010-10-31             415.79
2        12745.0          2     2010-08-10             723.85
3        12746.0          2     2010-06-30             266.35
4        12747.0         19     2010-12-13            4094.79
...          ...        ...            ...                ...
3884     18283.0          6     2010-11-22             641.77
3885     18284.0          2     2010-10-06             486.68
3886     18285.0          1     2010-02-17             427.00
3887     18286.0          2     2010-08-20             941.48
3888     18287.0          4     2010-11-22            2345.71

[3889 rows x 4 columns]

### Create a column for the number of days since the last order

In [9]:
today

Timestamp('2010-12-23 00:00:00')

In [8]:
today = df_clean['date'].max()
df_user['day_since_last_order'] = (today - df_user['max_order_date']).dt.days
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  
0                       80  
1                       53  
2                      135  
3                      176  
4                       10  
...                    ...  
3884                    31  
3885                    78  
3886                   309  
3887                   125  
3888                    31  

[3889 rows x 5 columns]

In [10]:
df_user.describe()

order_cnt  total_order_value  day_since_last_order
count  3889.000000        3889.000000           3889.000000
mean      5.128568        1544.623084             90.800720
std       8.499330        3434.816315             88.873286
min       1.000000           1.250000              0.000000
25%       1.000000         296.360000             25.000000
50%       3.000000         648.200000             58.000000
75%       6.000000        1585.940000            126.000000
max     163.000000       71970.390000            352.000000

### Create binning for the number of days since the last order, consisting of 5 bins with boundaries being the minimum, P20, P40, P60, P80, and maximum values. Assign labels 1 to 5 to the bins from highest to lowest as recency scores

In [11]:
df_user['recency_score'] = pd.cut(df_user['day_since_last_order'],
                                  bins=[df_user['day_since_last_order'].min(),
                                        np.percentile(df_user['day_since_last_order'], 20),
                                        np.percentile(df_user['day_since_last_order'], 40),
                                        np.percentile(df_user['day_since_last_order'], 60),
                                        np.percentile(df_user['day_since_last_order'], 80),
                                        df_user['day_since_last_order'].max()],
                                  labels=[5, 4, 3, 2, 1],
                                  include_lowest=True).astype(int)
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  recency_score  
0                       80              2  
1                       53              3  
2                      135              2  
3                      176              1  
4                       10              5  
...                    ...            ...  
3884                    31              4  
3885                    78              2  
3886                   309              1  
3887                   125              2  
3888                    31              4  

[3889 rows x 6 columns]

In [12]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3889 entries, 0 to 3888
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   customer_id           3889 non-null   object        
 1   order_cnt             3889 non-null   int64         
 2   max_order_date        3889 non-null   datetime64[ns]
 3   total_order_value     3889 non-null   float64       
 4   day_since_last_order  3889 non-null   int64         
 5   recency_score         3889 non-null   int32         
dtypes: datetime64[ns](1), float64(1), int32(1), int64(2), object(1)
memory usage: 167.2+ KB


### Create binning for the total number of transactions (orders), consisting of 5 bins with boundaries being the minimum, P20, P40, P60, P80, and maximum values. Assign labels 1 to 5 to the bins from lowest to highest as frequency scores

In [13]:
df_user['frequency_score'] = pd.cut(df_user['order_cnt'],
                                    bins=[0,
                                          np.percentile(df_user['order_cnt'], 20),
                                          np.percentile(df_user['order_cnt'], 40),
                                          np.percentile(df_user['order_cnt'], 60),
                                          np.percentile(df_user['order_cnt'], 80),
                                          df_user['order_cnt'].max()],
                                    labels=[1, 2, 3, 4, 5],
                                    include_lowest=True).astype(int)
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  recency_score  frequency_score  
0                       80              2                4  
1                       53              3                1  
2                      135              2                2  
3                      176              1                2  
4                       10              5                5  
...                    ...            ...              ...  
3884                    31              4                4  
3885                    78              2                2  
3886                   309              1                1  
3887                   125              2                2  
3888                    31              4                3  

[3889 rows x 7 columns]

### Create binning for the total order value, consisting of 5 bins with boundaries being the minimum, P20, P40, P60, P80, and maximum values. Assign labels 1 to 5 to the bins from lowest to highest as monetary scores

In [14]:
df_user['monetary_score'] = pd.cut(df_user['total_order_value'],
                                   bins=[df_user['total_order_value'].min(),
                                         np.percentile(df_user['total_order_value'], 20),
                                         np.percentile(df_user['total_order_value'], 40),
                                         np.percentile(df_user['total_order_value'], 60),
                                         np.percentile(df_user['total_order_value'], 80),
                                         df_user['total_order_value'].max()],
                                   labels=[1, 2, 3, 4, 5],
                                   include_lowest=True).astype(int)
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  recency_score  frequency_score  monetary_score  
0                       80              2                4               3  
1                       53              3                1               2  
2                      135              2                2               3  
3                      176              1                2               2  
4                       10              5                5               5  
...                    ...            ...              ...             ...  
3884                    31              4                4               3  
3885                    78              2                2               3  
3886                   309              1                1               2  
3887                   125              2                2               4  
3888                    31              4                3               5  

[3889 rows x 8 columns]

### Create a column for segment names based on recency and frequency scores

In [15]:
df_user['segment'] = np.select(
    [(df_user['recency_score']==5) & (df_user['frequency_score']>=4),
     (df_user['recency_score'].between(3, 4)) & (df_user['frequency_score']>=4),
     (df_user['recency_score']>=4) & (df_user['frequency_score'].between(2, 3)),
     (df_user['recency_score']<=2) & (df_user['frequency_score']==5),
     (df_user['recency_score']==3) & (df_user['frequency_score']==3),
     (df_user['recency_score']==5) & (df_user['frequency_score']==1),
     (df_user['recency_score']==4) & (df_user['frequency_score']==1),
     (df_user['recency_score']<=2) & (df_user['frequency_score'].between(3, 4)),
     (df_user['recency_score']==3) & (df_user['frequency_score']<=2),
     (df_user['recency_score']<=2) & (df_user['frequency_score']<=2)],
    ['01-Champion', '02-Loyal Customers', '03-Potential Loyalists', "04-Can't Lose Them", '05-Need Attention',
     '06-New Customers', '07-Promising', '08-At Risk', '09-About to Sleep', '10-Hibernating']
)
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  recency_score  frequency_score  monetary_score  \
0                       80              2                4               3   
1                       53              3                1               2   
2                      135              2                2               3   
3                      176              1                2               2   
4                       10              5                5               5   
...                    ...            ...              ...             ...   
3884                    31              4                4               3   
3885                    78              2                2               3   
3886                   309              1                1               2   
3887                   125              2                2               4   
3888                    31              4                3               5   

                     segment  
0                 08-At Risk  
1          09-About to Sleep  
2             10-Hibernating  
3             10-Hibernating  
4                01-Champion  
...                      ...  
3884      02-Loyal Customers  
3885          10-Hibernating  
3886          10-Hibernating  
3887          10-Hibernating  
3888  03-Potential Loyalists  

[3889 rows x 9 columns]

### Display a summary of the RFM segmentation (point 8) including the count, mean, and median of total orders, total order value, and number of days since the last order

In [16]:
summary = pd.pivot_table(df_user, index='segment',
               values=['customer_id','day_since_last_order','order_cnt','total_order_value'],
               aggfunc={'customer_id': pd.Series.nunique,
                        'day_since_last_order': [np.mean, np.median],
                        'order_cnt': [np.mean, np.median],
                        'total_order_value': [np.mean, np.median]})
summary['pct_unique'] = (summary['customer_id'] / summary['customer_id'].sum() * 100).round(1)
summary

customer_id day_since_last_order         order_cnt  \
                           nunique                 mean median       mean   
segment                                                                     
01-Champion                    550            10.618182    9.5  15.467273   
02-Loyal Customers             546            40.864469   37.0   8.767399   
03-Potential Loyalists         523            23.573614   24.0   2.829828   
04-Can't Lose Them              64           121.984375  112.5  11.375000   
05-Need Attention              176            58.613636   59.0   3.397727   
06-New Customers                50            14.220000   16.0   1.000000   
07-Promising                   142            32.760563   34.0   1.000000   
08-At Risk                     426           140.455399  120.0   4.136150   
09-About to Sleep              352            58.735795   58.0   1.417614   
10-Hibernating                1060           196.837736  199.0   1.313208   

                              total_order_value           pct_unique  
                       median              mean    median             
segment                                                               
01-Champion              10.0       5003.674245  2775.525       14.1  
02-Loyal Customers        7.0       2622.817826  1946.850       14.0  
03-Potential Loyalists    3.0        766.769828   622.070       13.4  
04-Can't Lose Them        9.5       2839.948125  2268.405        1.6  
05-Need Attention         3.0        989.232676   826.370        4.5  
06-New Customers          1.0        244.689000   193.675        1.3  
07-Promising              1.0        287.800282   238.440        3.7  
08-At Risk                4.0       1153.825683   875.430       11.0  
09-About to Sleep         1.0        448.229688   334.755        9.1  
10-Hibernating            1.0        343.083842   257.005       27.3

# Insight

1. Most users are in the Hibernating (1060 or 27.3%), Champion (550 or 14.1%), and Loyal Customers (546 or 14.0%) segments.
2. Special programs that focus on the urgency of transactions for Loyal Customers (546 or 14.0%) can be created to make them transact again in the near future so that they can move up to the Champion segment.
3. Special programs that focus on the number of transactions for Potential Loyalists (523 or 13.4%) can be created to make them transact more frequently so that they can move up to the Champion segment.
4. Special programs for Hibernating (1060 or 27.3%) can be made to make them return to transactions even though they are frequent so that they can move up to the New Customers or even Potential Loyalist segments.